packages/iTwinUI-react/src/core/Tree/Tree.tsx

```diff
@@ -224,40 +245,173 @@ export const Tree = <T,>(props: TreeProps<T>) => {
+          `#${nodeId}`,
+        ) as HTMLElement;
+        nodeElement?.focus();
+        // Need to reset that if navigating with mouse and keyboard,
+        // e.g. going to the previous node with keyboard
+        // and then clicking the next node with mouse.
+        setScrollToIndex(undefined);
+      }
+    });
+  }, [scrollToIndex]);
+
+  const handleFocus = (event: React.FocusEvent) => {
+    if (treeRef.current?.contains(event.relatedTarget)) {
+      return;
+    }
+
+    const items = getFocusableNodes();
+    if (items.length > 0) {
+      items[focusedIndex.current]?.focus();
+    }
+  };
+
+  return (
+    <>
+      {enableVirtualization ? (
+        <VirtualizedTree
+          flatNodesList={flatNodesList}
+          itemRenderer={itemRenderer}
+          scrollToIndex={scrollToIndex}
+          onFocus={handleFocus}
+          onKeyDown={handleKeyDown}
+          ref={treeRef}
+          className={className}
+          style={style}
+          {...rest}
+        />
+      ) : (
+        <TreeElement
+          onKeyDown={handleKeyDown}
+          onFocus={handleFocus}
+          className={className}
+          style={style}
+          ref={treeRef}
+          {...rest}
+        >
+          {flatNodesList.map((_, i) => itemRenderer(i))}
+        </TreeElement>
+      )}
+    </>
   );
 };
 
+type TreeElementProps = {
+  children: React.ReactNode;
+  onKeyDown: React.KeyboardEventHandler<HTMLUListElement>;
+  onFocus: React.FocusEventHandler<HTMLUListElement>;
+} & Omit<CommonProps, 'title'>;
+
+const TreeElement = React.forwardRef(
+  (
+    { children, className, ...rest }: TreeElementProps,
+    ref: React.ForwardedRef<HTMLUListElement>,
+  ) => {
+    return (
+      <ul
+        className={cx('iui-tree', className)}
+        role='tree'
+        ref={ref}
+        tabIndex={0}
+        {...rest}
+      >
+        {children}
+      </ul>
+    );
+  },
+);
+
+type VirtualizedTreeProps<T> = {
+  flatNodesList: FlatNode<T>[];
+  itemRenderer: (index: number) => JSX.Element;
+  scrollToIndex?: number;
+  onKeyDown: React.KeyboardEventHandler<HTMLUListElement>;
+  onFocus: React.FocusEventHandler<HTMLUListElement>;
+} & StylingProps;
+
+// Having virtualized tree separately prevents from running all virtualization logic
+const VirtualizedTree = React.forwardRef(
+  <T,>(
+    {
+      flatNodesList,
+      itemRenderer,
+      scrollToIndex,
+      onKeyDown,
+      onFocus,
+      className,
+      style,
+      ...rest
+    }: VirtualizedTreeProps<T>,
+    ref: React.ForwardedRef<HTMLUListElement>,
+  ) => {
+    const { outerProps, innerProps, visibleChildren } = useVirtualization({
+      itemsLength: flatNodesList.length,
+      itemRenderer: itemRenderer,
+      scrollToIndex,
+    });
+
+    return (
+      <div
+        {...{
+          ...outerProps,
+          className: cx(className, outerProps.className),
+          style: { ...style, ...outerProps.style },
```

Comment:
I thought it should go on the outter div for easier styling because users would need to have some wrapper or something to target the top element e.g. `.tree-wrapper > div`.

Reply:
That's a good point, but the rest props go on the tree element so it's really confusing.

I think we should probably support both. What is the currently expected way for the user to target the tree element when they use virtualization? `.my-outer-class > *`? Then we should document that.

Reply:
Added some comments.